File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    // Write your first then lower case function here
    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream()
                    .filter(predicate)
                    .findFirst()
                    .map(String::toLowerCase);
    }

    // Write your say function here
    static record Sayer(String phrase) {
        Sayer and(String word) {
            return new Sayer(phrase + " " + word);
        }
    }
    
    public static Sayer say() {
        return new Sayer("");
    }

    public static Sayer say(String word) {
        return new Sayer(word);
    }
 
    // Write your line count function here
    public static long meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            return reader.lines()
                    .filter(line -> !line.isBlank() && !line.trim().startsWith("#"))
                    .count();
        } catch (java.io.FileNotFoundException e) {
            throw new java.io.FileNotFoundException("No such file: " + filename);
        } catch (IOException e) {
            throw new IOException(e);
        }
    }
}

// Write your Quaternion record class here
record Quaternion(double a, double b, double c, double d) {
    public static final Quaternion ZERO = new Quaternion(0.0, 0.0, 0.0, 0.0);
    public static final Quaternion I = new Quaternion(0.0, 1.0, 0.0, 0.0);
    public static final Quaternion J = new Quaternion(0.0, 0.0, 1.0, 0.0);
    public static final Quaternion K = new Quaternion(0.0, 0.0, 0.0, 1.0);

    public Quaternion {
        if(Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    public Quaternion plus(Quaternion other) {
        return new Quaternion(this.a + other.a, this.b + other.b, this.c + other.c, this.d + other.d);
    }

    public Quaternion times(Quaternion other) {
        double q1 = this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d;
        double q2 = this.a * other.b + this.b * other.a + this.c * other.d - this.d * other.c;
        double q3 = this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b;
        double q4 = this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a;

        return new Quaternion(q1, q2, q3, q4);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    @Override
    public String toString() {
        List<String> parts = List.of(
            processCoefficient(this.a, ""), 
            processCoefficient(this.b, "i"), 
            processCoefficient(this.c, "j"), 
            processCoefficient(this.d, "k"));
            
        String result = String.join("", parts);
        if (!result.isEmpty() && result.charAt(0) == '+') {
            result = result.substring(1);
        }
        if (result.equals("")) {
            return "0";
        }
        return result;
    }

    private String processCoefficient(double coefficient, String letter) {
        if (coefficient == 0) {
            return "";
        } else if (coefficient == 1) {
            if (letter != "") {
                return "+" + letter;
            }
            return "+" + Double.toString(coefficient);
        } else if (coefficient == -1) {
            if (letter != "") {
                return "-" + letter;
            }
            return Double.toString(coefficient);
        } else if (coefficient > 0) {
            return "+" + Double.toString(coefficient) + letter;
        } else {
            return Double.toString(coefficient) + letter;
        }
    }
}

// Write your BinarySearchTree sealed interface and its implementations here
sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String value);
    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this); 
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        return "(" + (left instanceof Empty ? "" : left) + value + (right instanceof Empty ? "" : right) + ")";
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// Write your first then lower case function here
fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    return strings.firstOrNull(predicate)?.lowercase()
}

// Write your say function here
fun say(phrase: String = ""): Say {
    return Say(phrase)
}

data class Say(val phrase: String) {
    fun and(nextPhrase: String): Say {
        return Say("$phrase $nextPhrase")
    }
}

// Write your meaningfulLineCount function here
@Throws(IOException::class)
fun meaningfulLineCount(filename: String): Long {
    BufferedReader(FileReader(filename)).use { reader ->
        return reader.lines()
            .filter { line -> line.trim().isNotEmpty() && !line.trim().startsWith("#") }
            .count()
    }
}

// Write your Quaternion data class here
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion {
        val q1 = a * other.a - b * other.b - c * other.c - d * other.d
        val q2 = a * other.b + b * other.a + c * other.d - d * other.c
        val q3 = a * other.c - b * other.d + c * other.a + d * other.b
        val q4 = a * other.d + b * other.c - c * other.b + d * other.a
        return Quaternion(q1, q2, q3, q4)
    }

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

    override fun toString(): String {
        val parts = listOf(
            processCoefficient(a, ""),
            processCoefficient(b, "i"),
            processCoefficient(c, "j"),
            processCoefficient(d, "k")
        )
        val result = parts.joinToString("")
        return when {
            result.isNotEmpty() && result[0] == '+' -> result.substring(1)
            result.isEmpty() -> "0"
            else -> result
        }
    }

    private fun processCoefficient(coefficient: Double, letter: String): String {
        return when {
            coefficient == 0.0 -> ""
            coefficient == 1.0 && letter.isNotEmpty() -> "+$letter"
            coefficient == -1.0 && letter.isNotEmpty() -> "-$letter"
            coefficient > 0 -> "+${coefficient}${letter}"
            else -> "${coefficient}${letter}"
        }
    }
}

// Write your Binary Search Tree interface and implementing classes here
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }

    data class Node(
        private val value: String,
        private val left: BinarySearchTree,
        private val right: BinarySearchTree
    ) : BinarySearchTree {

        override fun size(): Int = 1 + left.size() + right.size()

        override fun contains(value: String): Boolean =  when {
            value < this.value -> left.contains(value)
            value > this.value -> right.contains(value)
            else -> true
        }

        override fun insert(value: String): BinarySearchTree = when {
            value < this.value -> Node(this.value, left.insert(value), right)
            value > this.value -> Node(this.value, left, right.insert(value))
            else -> this
        }

        override fun toString(): String {
            val result = StringBuilder()

            if (left !is Empty) {
                result.append(left.toString())
            }

            result.append(value)

            if (right !is Empty) {
                result.append(right.toString())
            }

            return "(${result})"
        }
    }
}